[JSC] Decode local indices at runtime in WASM in-place interpreter

ddegazio opened the pull request:
#### bfe0170022760b1772ed3378d20af384ca44b23d
<pre>
[JSC] Decode local indices at runtime in WASM in-place interpreter
<a href="https://bugs.webkit.org/show_bug.cgi?id=270376">https://bugs.webkit.org/show_bug.cgi?id=270376</a>
<a href="https://rdar.apple.com/123720682">rdar://123720682</a>

Reviewed by Justin Michaud and Yusuke Suzuki.

Removes metadata generation for local.get, local.set, and local.tee
from the WASM in-place interpreter. Instead, local indices for these
instructions are decoded at runtime. Since most local indices are
small, this actually saves us time - the added microbenchmark runs
3% faster with this change. More importantly, this saves on memory,
eliminating metadata for the most common WASM instructions.

* JSTests/microbenchmarks/local-get-set-tee.wasm: Added.
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::getLocal):
(JSC::Wasm::IPIntGenerator::setLocal):

Canonical link: <a href="https://commits.webkit.org/275815@main">https://commits.webkit.org/275815@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6af5798917a570027789d4feda5d95a10baa5591

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42906 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21927 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45307 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45524 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39031 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25628 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19307 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35506 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43479 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18953 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36945 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16479 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16561 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38002 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/962 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/36362 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39083 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38337 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47047 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/42533 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17733 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14620 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42261 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19358 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40917 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19522 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/49543 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5813 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18988 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10008 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -378,6 +378,11 @@ macro instructionLabel(instrname)
     _ipint%instrname%_validate:
 end
 
+macro slowPathLabel(instrname)
```
Does a slow path need to be aligned?

ddegazio replied:
Pretty sure it doesn't, since we never dispatch to it, it's just part of the normal `local.get/set/tee` implementation we're moving out-of-line since it's a rare case.

justinmichaud reviewed (approved):
r=me once EWS passes

Constellation reviewed (approved):
r=me

ddegazio committed:
[JSC] Decode local indices at runtime in WASM in-place interpreter
https://bugs.webkit.org/show_bug.cgi?id=270376
rdar://123720682

Reviewed by Justin Michaud and Yusuke Suzuki.

Removes metadata generation for local.get, local.set, and local.tee
from the WASM in-place interpreter. Instead, local indices for these
instructions are decoded at runtime. Since most local indices are
small, this actually saves us time - the added microbenchmark runs
3% faster with this change. More importantly, this saves on memory,
eliminating metadata for the most common WASM instructions.

* JSTests/microbenchmarks/local-get-set-tee.wasm: Added.
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::getLocal):
(JSC::Wasm::IPIntGenerator::setLocal):

Canonical link: https://commits.webkit.org/275815@main
